File: src/main/java/xyz/d1snin/corby/Corby.java
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2021, Corby
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package xyz.d1snin.corby;

import lombok.Getter;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.api.sharding.ShardManager;
import net.dv8tion.jda.api.utils.cache.CacheFlag;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import xyz.d1snin.corby.commands.Command;
import xyz.d1snin.corby.commands.admin.TerminateCommand;
import xyz.d1snin.corby.commands.fun.BottomCommand;
import xyz.d1snin.corby.commands.fun.CatCommand;
import xyz.d1snin.corby.commands.fun.UrbanCommand;
import xyz.d1snin.corby.commands.misc.HelpCommand;
import xyz.d1snin.corby.commands.misc.PingCommand;
import xyz.d1snin.corby.commands.misc.StealCommand;
import xyz.d1snin.corby.commands.misc.UptimeCommand;
import xyz.d1snin.corby.commands.settings.PrefixCommand;
import xyz.d1snin.corby.commands.settings.StarboardCommand;
import xyz.d1snin.corby.event.ReactionUpdateEvent;
import xyz.d1snin.corby.event.ServerJoinEvent;
import xyz.d1snin.corby.event.reactions.StarboardReactionEvent;
import xyz.d1snin.corby.manager.ConfigManager;
import xyz.d1snin.corby.manager.CooldownsManager;
import xyz.d1snin.corby.manager.LaunchArgumentsManager;
import xyz.d1snin.corby.model.Config;
import xyz.d1snin.corby.model.LaunchArgument;
import xyz.d1snin.corby.utils.FormatUtils;

import javax.security.auth.login.LoginException;
import java.awt.*;
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.lang.management.RuntimeMXBean;
import java.text.DecimalFormat;
import java.util.List;
import java.util.*;
import java.util.concurrent.Executors;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class Corby {

  @Getter private static Logger log;
  @Getter private static RuntimeMXBean rb;
  @Getter private static ForkJoinPool service;
  @Getter private static ScheduledExecutorService scheduler;
  @Getter private static DecimalFormat format;
  @Getter private static Random random;
  @Getter private static ShardManager shards;
  @Getter private static JDA firstJda;
  @Getter private static Set<Permission> permissions;
  @Getter private static List<Permission> defaultPermissions;
  @Getter private static List<String> presences;
  @Getter private static Config config;
  @Getter private static boolean testMode;
  @Getter private static boolean noShardsMode;

  public static void main(String[] args) {
    log = LoggerFactory.getLogger("loader");
    rb = ManagementFactory.getRuntimeMXBean();
    service = new ForkJoinPool();
    scheduler = Executors.newScheduledThreadPool(10);
    format = new DecimalFormat();
    random = new Random();
    permissions = new HashSet<>();
    presences = new ArrayList<>();
    testMode = false;
    noShardsMode = false;

    format.setDecimalSeparatorAlwaysShown(false);

    defaultPermissions =
        Arrays.asList(
            Permission.MESSAGE_HISTORY,
            Permission.MESSAGE_READ,
            Permission.MESSAGE_WRITE,
            Permission.MESSAGE_MANAGE,
            Permission.VIEW_CHANNEL,
            Permission.MANAGE_CHANNEL);

    try {
      LaunchArgumentsManager.init(
          args,
          new LaunchArgument(
              "test",
              () -> {
                testMode = true;
                log.warn("Start using the bot token for testing...");
              }),
          new LaunchArgument(
              "noshards",
              () -> {
                noShardsMode = true;
                log.warn("Starting without sharding...");
              }));

      log.info("Starting...");

      permissions.addAll(defaultPermissions);

      start();

      startUpdatePresence();
      CooldownsManager.startUpdating();

    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  public static void start() throws LoginException, IOException, InterruptedException {
    config = ConfigManager.init();

    shards =
        DefaultShardManagerBuilder.createDefault(
                testMode ? config.getTestBotToken() : config.getToken())
            .enableIntents(GatewayIntent.GUILD_PRESENCES, GatewayIntent.GUILD_MESSAGE_REACTIONS)
            .enableCache(
                CacheFlag.CLIENT_STATUS,
                CacheFlag.VOICE_STATE,
                CacheFlag.ACTIVITY,
                CacheFlag.ROLE_TAGS,
                CacheFlag.EMOTE)
            .setStatus(OnlineStatus.IDLE)
            .addEventListeners(
                new StarboardReactionEvent(), new ReactionUpdateEvent(), new ServerJoinEvent())
            .addEventListeners(
                new ArrayList<>(
                    Command.addAll(
                        // admin category
                        new TerminateCommand(),
                        // fun category
                        new BottomCommand(),
                        new CatCommand(),
                        new UrbanCommand(),
                        // misc category
                        new HelpCommand(),
                        new PingCommand(),
                        new StealCommand(),
                        new UptimeCommand(),
                        // bot settings category
                        new PrefixCommand(),
                        new StarboardCommand())))
            .setShardsTotal(isNoShardsMode() ? 1 : getConfig().getShardsTotal())
            .build();

    for (JDA jda : shards.getShards()) {
      jda.awaitReady();
    }

    System.out.println(
        "\n"
            + String.join(
                "\n",
                "   ██████╗ ██████╗ ██████╗ ██████╗ ██╗   ██╗  ",
                "  ██╔════╝██╔═══██╗██╔══██╗██╔══██╗╚██╗ ██╔╝  ",
                "  ██║     ██║   ██║██████╔╝██████╔╝ ╚████╔╝   ",
                "  ██║     ██║   ██║██╔══██╗██╔══██╗  ╚██╔╝    ",
                "  ╚██████╗╚██████╔╝██║  ██║██████╔╝   ██║     ",
                "   ╚═════╝ ╚═════╝ ╚═╝  ╚═╝╚═════╝    ╚═╝     ")
            + "\n");

    firstJda = shards.getShards().get(0);

    config.initOther(
        new Color(222, 222, 222),
        new Color(255, 75, 75),
        new Color(112, 228, 120),
        new Color(255, 215, 0),
        firstJda.getSelfUser().getName(),
        firstJda.getSelfUser().getEffectiveAvatarUrl(),
        firstJda.getInviteUrl(permissions),
        firstJda.getSelfUser().getId(),
        firstJda.getSelfUser().getAsTag());

    log = LoggerFactory.getLogger(config.getBotName());

    log.info(
        String.format(
            "Bot has started up in %s!\n"
                + "    ~ PFP:         %s\n"
                + "    ~ Name:        %s\n"
                + "    ~ ID:          %s\n"
                + "    ~ Invite URL:  %s\n",
            getUptime(),
            config.getBotPfpUrl(),
            config.getNameAsTag(),
            config.getId(),
            config.getInviteUrl()));

    if (!isNoShardsMode()) {
      log.warn("Shards loading can be long\n");
    }
  }

  private static void startUpdatePresence() {
    scheduler.scheduleWithFixedDelay(
        () -> shards.setActivity(Activity.watching(String.format(";help | %s", getPresence()))),
        0,
        7,
        TimeUnit.SECONDS);
  }

  private static String getPresence() {
    presences.clear();
    presences.add(String.format("Ping: %s", getPing()));
    presences.add(String.format("%d Servers!", shards.getGuilds().size()));
    return presences.get(random.nextInt(presences.size()));
  }

  public static void shutdown(int exitCode) {
    log.warn("Terminating... Bye!");
    shards.shutdown();
    scheduler.shutdown();
    service.shutdown();
    System.gc();
    System.exit(exitCode);
  }

  public static String getUptime() {
    return FormatUtils.formatMillis(rb.getUptime());
  }

  public static String getPing() {
    return format.format(getShards().getAverageGatewayPing());
  }
}

File: src/main/kotlin/xyz/d1snin/corby/Corby.kt
/*
 * BSD 3-Clause License, Copyright (c) 2021, Corby and contributors, All rights reserved.
 */

package xyz.d1snin.corby

import ch.qos.logback.classic.Level
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder
import net.dv8tion.jda.api.sharding.ShardManager
import net.dv8tion.jda.api.utils.cache.CacheFlag
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import xyz.d1snin.corby.commands.`fun`.BottomCommand
import xyz.d1snin.corby.commands.`fun`.UrbanCommand
import xyz.d1snin.corby.commands.admin.TerminateCommand
import xyz.d1snin.corby.commands.misc.HelpCommand
import xyz.d1snin.corby.commands.misc.PingCommand
import xyz.d1snin.corby.database.DatabaseManager
import xyz.d1snin.corby.event.ButtonEvent
import xyz.d1snin.corby.event.ServerJoinEvent
import xyz.d1snin.corby.event.reactions.StarboardReactionEvent
import xyz.d1snin.corby.manager.CommandsManager
import xyz.d1snin.corby.manager.CooldownsManager
import xyz.d1snin.corby.manager.ListenersManager
import xyz.d1snin.corby.util.Configs
import xyz.d1snin.corby.util.formatTimeMillis
import java.awt.Color
import java.io.File
import java.lang.management.ManagementFactory
import java.lang.management.RuntimeMXBean
import java.text.DecimalFormat
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit
import kotlin.system.exitProcess

object Corby {
    private const val CONFIG_FILE = "config.json"

    // exit codes
    const val GOOD_EXIT_CODE = 0
    const val DATABASE_ERROR = 10

    val defaultScope = CoroutineScope(Dispatchers.Default)

    lateinit var log: Logger
    lateinit var config: Configs
    lateinit var scheduler: ScheduledExecutorService
    lateinit var sharding: ShardManager
    lateinit var firstShard: JDA
    lateinit var permissions: MutableSet<Permission>
    lateinit var selfUser: User

    private lateinit var rb: RuntimeMXBean
    private lateinit var format: DecimalFormat
    private lateinit var defaultPermissions: Set<Permission>

    val ping: String
        get() = format.format(sharding.averageGatewayPing)

    val uptime: String
        get() = rb.uptime.formatTimeMillis()

    private val presence
        get() = listOf(
            "Ping: $ping",
            "${sharding.guilds.size} Servers!",
            "Uptime: $uptime"
        ).run {
            shuffled().first()
        }

    fun launch(args: Array<String>) {
        log = LoggerFactory.getLogger("Loader")

        log("Launching...")

        val arguments = Arguments().also {
            JCommander.newBuilder()
                .addObject(it)
                .build()
                .parse(*args)
        }

        defaultPermissions = setOf(
            Permission.MESSAGE_HISTORY,
            Permission.MESSAGE_READ,
            Permission.MESSAGE_WRITE,
            Permission.MESSAGE_MANAGE,
            Permission.VIEW_CHANNEL,
            Permission.MANAGE_CHANNEL
        )

        scheduler = ScheduledThreadPoolExecutor(10)
        permissions = mutableSetOf<Permission>().apply { addAll(defaultPermissions) }
        rb = ManagementFactory.getRuntimeMXBean()
        format = DecimalFormat()
        config = Configs.init(File(CONFIG_FILE))

        sharding = DefaultShardManagerBuilder.createDefault(
            if (arguments.isTestMode) config.testBotToken else config.token
        ).run {
            addEventListeners(
                ListenersManager.addAll(
                    StarboardReactionEvent,
                    ButtonEvent,
                    ServerJoinEvent
                )
            )

            addEventListeners(
                CommandsManager.addAll(
                    TerminateCommand,
                    BottomCommand,
                    UrbanCommand,
                    HelpCommand,
                    PingCommand
                )
            )

            enableIntents(
                GatewayIntent.GUILD_PRESENCES,
                GatewayIntent.GUILD_MESSAGE_REACTIONS
            )

            enableCache(
                CacheFlag.CLIENT_STATUS,
                CacheFlag.VOICE_STATE,
                CacheFlag.ACTIVITY,
                CacheFlag.ROLE_TAGS,
                CacheFlag.EMOTE
            )

            setStatus(OnlineStatus.IDLE)

            setShardsTotal(if (arguments.isNoShards) 1 else config.shards)

            build()
        }

        firstShard = sharding.shards.first()
        selfUser = firstShard.selfUser

        config.other(
            Color(222, 222, 222),
            Color(255, 75, 75),
            Color(112, 228, 120),
            Color(255, 215, 0),
            selfUser.name,
            selfUser.effectiveAvatarUrl,
            firstShard.getInviteUrl(permissions),
            selfUser.id,
            selfUser.asTag
        )

        DatabaseManager.init()
        CooldownsManager.startUpdating()
        startPresenceUpdating()

        log = LoggerFactory.getLogger(config.botName)

        log("Bot has started up in $uptime!")
    }

    internal fun shutdown(exitCode: Int) {
        log.warn("Terminating... Bye!")
        sharding.shutdown()
        scheduler.shutdown()
        exitProcess(exitCode)
    }

    fun log(message: String, level: Level = Level.INFO) = when (level) {
        Level.DEBUG -> log.debug(message)
        Level.ERROR -> log.error(message)
        Level.TRACE -> log.trace(message)
        Level.WARN -> log.warn(message)
        else -> log.info(message)
    }

    private fun startPresenceUpdating() {
        scheduler.scheduleWithFixedDelay(
            {
                sharding.setActivity(Activity.watching(presence))
            }, 0, 10, TimeUnit.SECONDS
        )
    }

    private class Arguments {
        @Parameter(
            names = ["--test", "-T"],
            description = "Launch the bot using token for testing."
        )
        var isTestMode = false

        @Parameter(
            names = ["--noshards", "-NS"],
            description = "Launch the bot without sharding."
        )
        var isNoShards = false
    }
}